gravitee-am-service/src/main/java/io/gravitee/am/service/impl/VerifyAttemptServiceImpl.java

```diff
@@ -0,0 +1,167 @@
+ * limitations under the License.
+ */
+package io.gravitee.am.service.impl;
+
+import io.gravitee.am.common.audit.EventType;
+import io.gravitee.am.model.Domain;
+import io.gravitee.am.model.ReferenceType;
+import io.gravitee.am.model.User;
+import io.gravitee.am.model.VerifyAttempt;
+import io.gravitee.am.model.account.AccountSettings;
+import io.gravitee.am.model.oidc.Client;
+import io.gravitee.am.repository.management.api.VerifyAttemptRepository;
+import io.gravitee.am.repository.management.api.search.VerifyAttemptCriteria;
+import io.gravitee.am.service.AuditService;
+import io.gravitee.am.service.EmailService;
+import io.gravitee.am.service.VerifyAttemptService;
+import io.gravitee.am.service.reporter.builder.AuditBuilder;
+import io.gravitee.am.service.reporter.builder.management.VerifyAttemptAuditBuilder;
+import io.reactivex.Completable;
+import io.reactivex.Maybe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+import java.util.Optional;
+
+/**
+ * @author Ashraful Hasan (ashraful.hasan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@Component
+public class VerifyAttemptServiceImpl implements VerifyAttemptService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(VerifyAttemptServiceImpl.class);
+
+    @Autowired
+    AuditService auditService;
+
+    @Autowired
+    EmailService emailService;
+
+    @Lazy
+    @Autowired
+    VerifyAttemptRepository verifyAttemptRepository;
+
+    @Override
+    public Maybe<VerifyAttempt> checkVerifyAttempt(String userId, String factorId, Client client, Domain domain) {
+        final AccountSettings accountSettings = AccountSettings.getInstance(domain, client);
+        if (accountSettings == null || !accountSettings.isMfaChallengeAttemptsDetectionEnabled()) {
+            LOGGER.debug("MFA brute force detection is disabled, won't check verify attempt.");
+            return Maybe.empty();
+        }
+
+        final VerifyAttemptCriteria criteria = buildCriteria(userId, factorId, client.getId());
+
+        return getVerifyAttemptIfExists(criteria, accountSettings)
+                .doOnSuccess(verifyAttempt -> {
+                    LOGGER.debug("VerifyAttempt value: [{}]", verifyAttempt);
+                    if (!verifyAttempt.isAllowRequest()) {
+                        auditService.report(AuditBuilder.builder(VerifyAttemptAuditBuilder.class)
+                                .type(EventType.MFA_VERIFICATION_LIMIT_EXCEED)
+                                .verifyAttempt(verifyAttempt));
+                    }
+                });
+    }
+
+    @Override
+    public Completable verificationFailed(String userId, String factorId, Client client, Domain domain, Optional<VerifyAttempt> optionalVerifyAttempt) {
```

Comment:
open discussion:

IMHO this method name doesn't reflect the purpose of the method.
`incrementAttempt` or something like that could be more explicit.

wdyt ?

Reply:
method renamed as suggested